engine/include/cubos/engine/ui/canvas/element.hpp

```diff
@@ -20,24 +23,32 @@ namespace cubos::engine
     struct CUBOS_ENGINE_API UIElement
     {
         CUBOS_REFLECT;
-
         glm::vec2 offset = {0.0F, 0.0F}; ///< Offset of the pivot.
         glm::vec2 size = {0.0F, 0.0F};   ///< Size of the element.
         glm::vec2 pivot = {0.5F, 0.5F};  ///< Relative position of the pivot to the element itself.
         glm::vec2 anchor = {0.5F, 0.5F}; ///< Relative position of the pivot to the parent element.
         int layer = 0;                   ///< Rendering order layer.
 
         glm::vec2 position = {0.0F, 0.0F}; ///< Global position of the pivot.
-        glm::mat4 vp = {};                 ///< View matrix provided by the canvas.
 
         int hierarchyDepth = 0; ///< Depth in hierarchy. Used for depth sorting.
 
+        UIDrawList drawList; ///< Draw list holding draw commands from this element.
+
         /// @brief Gets global position of the bottom left corner of the element.
         /// @return Position value in global coordinates.
         glm::vec2 bottomLeftCorner() const;
 
         /// @brief Gets global position of the top right corner of the element.
         /// @return Position value in global coordinates.
         glm::vec2 topRightCorner() const;
+
+        /// @brief Adds entry to draw list.
+        /// @param entry Entry to add.
+        /// @param data Pointer to per element data.
+        void pushEntry(UIDrawListEntry entry, const void* data);
```

Comment:
Wouldn't it be more friendly to return a builder here to modify the added entry?
Since this is what the user will interact with when making custom widgets and stuff it should look as good as possible.
I.e. `Builder draw(const UIDrawType& type, core::gl::VertexArray vertexArray, size_t vertexOffset, size_t vertexCount, const void* data)`

This was what I was suggesting before, sorry for not making it clearer
What I was imagining was being able to do something like:
```
element.draw(type, va, 0, 6, perElementData).withTexture(0, myTexture);
```
Basically the builder would just be a wrapper around a reference to an UIDrawEntry.
What do you think?

---

As a sidenote, I would maybe name this `drawUntyped`, and add a wrapper templated method which calls it. For example,

```cpp
template <typename T>
Builder draw(const UIDrawType& type, core::gl::VertexArray vertexArray, size_t vertexOffset, size_t vertexCount, const T& data)
{
  return this->drawUntyped(type, vertexArray, vertexOffset, vertexCount, &data);
}
```

Then the color rect code could become just:

```cpp
element.draw(state.drawType, state.va, 0, 6, PerElement{
  .xRange = {minX, maxX},
  .yRange = {minY, maxY},
  .color = colorRect.color,
  .depth = element.layer * 100 + element.hierarchyDepth
});
```